### ISSUE TITLE
dask array failing with numpy 1.17.0rc1 and np.min

### ISSUE THREAD

**Issue by jni** [dask/dask#5031]
Using Python 3.7, dask 2.0, (from conda-forge) and NumPy 1.17.0rc1 (from pip):

```python
import numpy as np
import dask.array as da

image = da.from_array(np.array([[0, 1], [1, 2]]), chunks=(1, 2))
image_min = int(np.min(image))
```

results in:

```pytb
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 6, in amin
TypeError: no implementation found for 'numpy.amin' on types that implement __array_function__: [<class 'dask.array.core.Array'>]
```

**Comment by jni** [comment]
xref https://github.com/scikit-image/scikit-image/issues/3985

**Comment by shoyer** [comment]
It looks like NumPy defines `numpy.min` as an alias of `numpy.amin`. `dask.array.min` exists but not `dask.array.amin`, so dask's `__array_function__` method doesn't pick it up.

**Comment by rgommers** [comment]
There'll be many of these though. 
```
In [1]: import dask.array as da                                                                

In [2]: x = np.arange(5)                                                                       

In [3]: d = da.from_array(x)                                                                   

In [4]: np.__version__                                                                         
Out[4]: '1.16.4'

In [5]: np.median(d)                                                                           
Out[5]: 2.0
```

And with 1.17.0 it doesn't work anymore:
```
In [1]: import dask.array as da                                                                

In [2]: x = np.arange(5)                                                                       

In [3]: d = da.from_array(x)                                                                   

In [4]: np.__version__                                                                         
Out[4]: '1.17.0.dev0+f0e64a1'

In [5]: np.median(d)                                                                           
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-4bf961ea5541> in <module>
----> 1 np.median(d)

<__array_function__ internals> in median(*args, **kwargs)

TypeError: no implementation found for 'numpy.median' on types that implement __array_function__: [<class 'dask.array.core.Array'>]
```

Anyone who was relying on NumPy functions coercing to ndarray is now going to have to do that explicity
```
In [6]: np.median(np.asarray(d))                                                               
Out[6]: 2.0
```

It's actually a quite significantly backwards compat break, without any deprecation warnings in the previous release ....

**Comment by rgommers** [comment]
It may not be too late to change `TypeError` with a warning + old behavior.

**Comment by eric-wieser** [comment]
To more concretely summarize the problem here, since I was confused by the mailing list post - this is `dask`'s `__array_function__` implementation:

https://github.com/dask/dask/blob/a57b02992be979abfaa347339540f1b20cc6ff28/dask/array/core.py#L1208-L1221

The problem is that the function `__name__` is being used to map between numpy and dask functions, which makes dask sensitive to which function is the original and which is the alias. This, for instance, makes swapping `np.min` and `np.amin` to be aliases in the other direction a breaking change.

In my opinion, we should document that `np.some_func.__name__` is an implementation detail, and not something that should be relied upon inside `__array_ufunc__`. Users should instead create a `dict` mapping from numpy to their functions, perhaps with a decorator like `@matches_numpy_api(np.abs)`, to make it easy to do so.

**Comment by rgommers** [comment]
I like the dict mapping idea, that deal with aliases and name mismatches nicely. Don't think I like the decorator so much, that messes with stacklevel and isn't less typing than an explicit mapping.

**Comment by shoyer** [comment]
> It's actually a quite significantly backwards compat break, without any deprecation warnings in the previous release ....

To be clear, this a backwards compatibility break within *dask*, not NumPy. But I agree, for libraries like dask that also implement `__array__` it would be friendlier to define `__array_function__` for unimplemented functions by coercing arguments to NumPy arrays and then calling the NumPy function again. Something like:
```
     if not hasattr(module, func.__name__): 
         warning.warn('not implemented by dask')
         # inspect args/kwargs to find/replace dask arrays with their
         # computed versions
         return func(*computed_args, **computed_kwargs)
```

**Comment by eric-wieser** [comment]
> Don't think I like the decorator so much, that messes with stacklevel

That's not true:
```python
def matches_numpy_api(np_func):
    def decorator(my_func):
        _numpy_to_my_funcs[np_func] = my_func
        return my_func  # no wrapper created
    return decorator
```

**Comment by shoyer** [comment]
The decorator giving explicit aliases is already the recommendation from
the NEP. But for a library like dask that already maintains a module with
an API matching NumPy, matching names is definitely much easier. And it
works already, probably 98% of the time.

**Comment by jni** [comment]
> But I agree, for libraries like dask that also implement `__array__` it would be friendlier to define `__array_function__` for unimplemented functions by coercing arguments to NumPy arrays and then calling the NumPy function again.

:point_up: this! This enables gradual implementation of `__array_function__`, rather than having it be all-or-nothing. And as @shoyer pointed out in the scikit-image issue, this is already what many libraries rely on with dask arrays anyway: transparent coercion to NumPy arrays, which will work fine in cases where dask is used for multiprocessing rather than for out-of-core/distributed computations.

Having said this, the min/amin problem is a big one and will hopefully be addressed by the dict/decorator approach.

> To be clear, this a backwards compatibility break within dask, not NumPy.

Well, yes and no? If `__array_function__` was considered experimental, then it's somewhat up to NumPy not to break partial implementations of it?

**Comment by pentschev** [comment]
> Well, yes and no? If `__array_function__` was considered experimental, then it's somewhat up to NumPy not to break partial implementations of it?

To me, experimental means that it isn't guaranteed no breaking changes will happen, since there may be issues with the original proposal or new features added that were deemed necessary after some experimental testing.

Thanks @mrocklin for the fix.

**Comment by rgommers** [comment]
Fix is in master now. That still leaves the issue of what to do with latest released dask + numpy 1.17.0. At the moment that's still going to break scikit-image (discussion in https://github.com/scikit-image/scikit-image/issues/3985). Would a bugfix release be an option @mrocklin, so there's at least one compatible dask/numpy version combination that people could upgrade to?

**Comment by jrbourbeau** [comment]
> Would a bugfix release be an option

+1 from me (ref #5063)

**Comment by jakirkham** [comment]
We had previously discussed doing a bug fix soonish. Were there other fixes we needed?

**Comment by jrbourbeau** [comment]
Closing as Dask 2.1.0 is out which includes the fix from #5043. Thanks all!
